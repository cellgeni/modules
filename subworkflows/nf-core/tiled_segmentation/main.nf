include { TILEDIMAGESEGMENTATION_CELLPOSEWRAPPER as CELLPOSE } from '../../../modules/nf-core/tiledimagesegmentation/cellposewrapper'
include { TILEDIMAGESEGMENTATION_STARDISTWRAPPER as STARDIST} from '../../../modules/nf-core/tiledimagesegmentation/stardistwrapper'
include { TILEDIMAGESEGMENTATION_INSTANSEGWRAPPER as INSTANSEG} from '../../../modules/nf-core/tiledimagesegmentation/instansegwrapper'
include { TILEDIMAGESEGMENTATION_DEEPCELLWRAPPER as DEEPCELL} from '../../../modules/nf-core/tiledimagesegmentation/deepcellwrapper'
include { TILEDIMAGESEGMENTATION_MERGEOUTLINES as MERGEOUTLINES} from '../../../modules/nf-core/tiledimagesegmentation/mergeoutlines'
include { TILEDIMAGESEGMENTATION_GENERATETILECOORDS as GENERATE_TILE_COORDS } from '../../../modules/nf-core/tiledimagesegmentation/generatetilecoords'


workflow TILED_SEGMENTATION {
    take:
    images 
    method

    main:
    ch_versions = Channel.empty()
    GENERATE_TILE_COORDS(images)
    ch_versions = ch_versions.mix(GENERATE_TILE_COORDS.out.versions.first())

    images_tiles = GENERATE_TILE_COORDS.out.tile_coords.splitCsv(header:true, sep:",").map{ meta, coords ->
        [meta, coords.X_MIN, coords.Y_MIN, coords.X_MAX, coords.Y_MAX]
    }
    tiles_and_images = images_tiles.combine(images, by:0)
    if (method == "CELLPOSE") {
        CELLPOSE(tiles_and_images.combine(channel.from(params.cell_diameters)))
        wkts = CELLPOSE.out.wkts.groupTuple(by:0)
        ch_versions = ch_versions.mix(CELLPOSE.out.versions.first())
    } else if (method == "STARDIST") {
        STARDIST(tiles_and_images)
        wkts = STARDIST.out.wkts.groupTuple(by:0)
        ch_versions = ch_versions.mix(STARDIST.out.versions.first())
    } else if (method == "INSTANSEG") {
        INSTANSEG(tiles_and_images)
        wkts = INSTANSEG.out.wkts.groupTuple(by:0)
        ch_versions = ch_versions.mix(INSTANSEG.out.versions.first())
    } else if (method == "DEEPCELL") {
        DEEPCELL(tiles_and_images)
        wkts = DEEPCELL.out.wkts.groupTuple(by:0)
        ch_versions = ch_versions.mix(DEEPCELL.out.versions.first())
    } else {
        error "Invalid segmentation method: ${method}. Expected one of: CELLPOSE, STARDIST, INSTANSEG, DEEPCELL"
    }
    MERGEOUTLINES(wkts)
    ch_versions = ch_versions.mix(MERGEOUTLINES.out.versions.first())

    emit:
    geojson         = MERGEOUTLINES.out.multipoly_geojsons   // channel: [ val(meta), [ geojson ] ]
    versions    = ch_versions                     // channel: [ versions.yml ]
}