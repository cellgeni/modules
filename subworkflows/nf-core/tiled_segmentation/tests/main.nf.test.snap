{
    "nuclear - tif - instanseg": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,50ed2ca89950dc384bdc09889962dfc2"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b",
                    "versions.yml:md5,74ae9680e8565f16244a79acbe2ef1e4"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,50ed2ca89950dc384bdc09889962dfc2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b",
                    "versions.yml:md5,74ae9680e8565f16244a79acbe2ef1e4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-05-06T23:08:16.92761512"
    },
    "nuclear - tif - cellpose": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,3960f226fb5ded5fefab641613ee3f12"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,1eaa08f3c3bfc2b6487998c5870c6820",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,3960f226fb5ded5fefab641613ee3f12"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,1eaa08f3c3bfc2b6487998c5870c6820",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-05-06T23:07:12.303577971"
    },
    "nuclear - tif - deepcell": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,bc2c1795bcb9c985641d97e0f9d8477c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b",
                    "versions.yml:md5,e693a7e9c66cd7a37d9829cae2fc3648"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,bc2c1795bcb9c985641d97e0f9d8477c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b",
                    "versions.yml:md5,e693a7e9c66cd7a37d9829cae2fc3648"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-05-06T23:09:38.362846834"
    },
    "nuclear - tif - stardist": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,590cf1b456fbae16631254f0f79121fc"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,305c2355771c25305b764a1be9b39358",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b"
                ],
                "geojson": [
                    [
                        {
                            "id": "test"
                        },
                        "test_merged.geojson:md5,590cf1b456fbae16631254f0f79121fc"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d6e2b331dc05b3496a477c7e8adbeee",
                    "versions.yml:md5,305c2355771c25305b764a1be9b39358",
                    "versions.yml:md5,4859ae13b542b139121d9feca296312b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-05-06T23:07:45.544643239"
    }
}